AEXS/aex-141.md

```diff
@@ -393,6 +403,108 @@ stateful entrypoint burn(token_id: int) : unit
 | :--- | :--- |
 | token_id | int |
 
+## Extension Mintable Templates ("mintable_templates")
+
+The `mintable_templates` extension SHOULD be used if multiple NFTs within a contract should share the same metadata. The maximum edition size of NFTs for a specific template needs to be defined on template creation.
+
+```sophia
+contract interface IAEX141MintableTemplates =
+
+    record template =
+        { immutable_metadata_url: string
+        , edition_size: int
+        , edition_mint_count: int }
```

Comment:
There are use-cases where the creator might want to lower or extend the collection size (`edition_size`) at runtime (see most of the drops in the Ethereum space).

I suggest this template to also provide entrypoints in its interface for modifying those by the creator.

Reply:
hey @mradkov, nice to get feedback from your side! made me smile today 😃 

can you provide me a concrete example? I generally agree with lowering the `edition_size` because one might want to decrease the max amount afterwards.

I am a bit concerned and worrying about increasing the `edition_size` afterwards. at least users should have the possibility here to check if this is possible.

an example usecase would help me to evaluate that topic.

---------
regarding 
> collection size (`edition_size`) at runtime

I just want to be sure we talk about the same thing. as the standard here currently proposes each contract reflects ONE collection and each collection could have (potentially) unlimited amount of NFTs included. a cap of the max. amount of NFTs is currently not proposed and could of course be implemented individually.

the `mintable_templates` is about the edition size of a specific NFT-template (e.g. Pickachu) where the creator would e.g. define that max. 1000 Pickachus could exist. combined with `mutable_attributes` extension each unique Pickachu could then be leveled up individually.

personally as a user / collector I wouldn't be happy if the possible max edition size of the Pickachu would suddenly be increased to 10k instead of 1k. this is why I am asking 😅

Reply:
@mradkov any comment on the topic to increase the edition size after defining the template?

Reply:
For example Bored Apes Yacht Club - they have an infinite mint possibility by the owners of the contract - https://twitter.com/dafky2000/status/1400171615061741569. (they've said they destroyed their owner keys but you never know).

In my opinion the standard should not forbid such behaviour.

Reply:
> For example Bored Apes Yacht Club - they have an infinite mint possibility by the owners of the contract - https://twitter.com/dafky2000/status/1400171615061741569. (they've said they destroyed their owner keys but you never know).
> 
> In my opinion the standard should not forbid such behaviour but this should be part of the business logic a project has.

ok so this is about a limitation of the unique NFTs or the templates of an NFT collection, right? do you think this could be a separate extension (or even two extensions) or would you see these included in `mintable` / `mintable_templates` ?

I agree it would be good to have a standardized way to indicate this.

> The examples you gave is from the art/investment standpoint and I agree with those, but a NFT standard should not be limited only to those.
>
> Example: In an imaginary use-case in the gaming space - you may want to have unlimited in-game assets (weapons, land, character traits, NPCs, etc.) of the same kind possibly minted so that you ensure the longevity of your game.

I am really wondering if we need to cover the **increase** of the `edition_size` of a specific NFT template. I see two option for game providers:
- use `0` edition size for each template to be flexible
- create new templates or even new collections (e.g. series 2 of a specific item)

personally I think even in games people would want to be sure their items are rare. what's the meaning of an edition size if it can be altered and increased later on? frontends can always show the current circulating supply.

I would definitely add sth. to lower the edition size of a template afterwards. but I am still **very** worried about increasing it. plus this adds (IMO) unnecessary complexity

Reply:
other opinions? @arjanvaneersel @jyeshe

Reply:
I 100% agree with you @mradkov and I highly appreciate your feedback. except the increase of max cap / edition size I would actually take everything into consideration :)

- max cap (including lowering the max cap)
   - the question is where to include 🤔 currently I think it's best placed in the respective extension `mintable` or `mintable_templates`
- lowering edition size of templates in `mintable_templates` extension

what's your take on where to place the max cap stuff?

Reply:
> - max cap (including 0 for unlimited and possibility to lower the max cap afterwards)
>   - the question is where to include 🤔 currently I think it's best placed in the respective extension mintable or mintable_templates
> - lowering edition size of templates in mintable_templates extension
> 
> what's your take on where to place the max cap stuff?
> 

In my opinion it will be best if all of the extensions are composable so that different sets of business requirements can be met by just combining a bunch of extensions. It's also safer and less error prone that way.

So a separate composable extension make sense to me.

Reply:
reference implementation now updated. AEX standard document still needs to be updated here. for the implementation check out:
- `mintable`, `mintable_limit` & `burnable` -> https://github.com/aeternity/aex141-nft-collection-example/blob/master/contracts/CollectionUniqueNFTs.aes
- `mintable_templates`, `mintable_templates_limit`, `mutable_attributes` & `burnable` -> https://github.com/aeternity/aex141-nft-collection-example/blob/master/contracts/CollectionTemplateEditionNFTs.aes

Reply:
In order to simplify the standard and examples, maybe the limit related extensions could be a separate folder. It would be a good opportunity to open/start a place for community extensions.

Reply:
we need to make use of polymorphism feature when it's finally supported by SDK and http compiler v7 is deployed. will updated the examples then. I agree it's very complicated to combine various extension together right now 😅 

can we resolve this conversation?

Reply:
resolving this conversation now. from my perspective it's all covered now (finally). but we only allow decreasing the edition limit and template limit